From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Marcus Hall <marcus@tuells.org>
Date: Wed, 27 Jan 2021 13:43:10 -0700
Subject: [PATCH] 2695: CALL_RETURN override of non-computed conditional

With CALL_RETURN override, if the underlying flow is conditional but not computed, it should return CONDITIONAL_CALL_TERMINATOR.  Previously if not conditional but terminal it returned COMPUTED_CALL_TERMINATOR which is clearly wrong since it isn't computed and otherwise the override didn't modify the flow at all.  I got here looking at a conditional branch to a subroutine which should be a tail optimized branch to a call and return, so with CALL_RETURN override applied I expect it to be CONDITIONAL_CALL_TERMINATOR.
---
 .../java/ghidra/program/model/listing/FlowOverride.java     | 6 +-----
 1 file changed, 1 insertion(+), 5 deletions(-)

diff --git a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/listing/FlowOverride.java b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/listing/FlowOverride.java
index 8cf525f2d9..65dc673fed 100644
--- a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/listing/FlowOverride.java
+++ b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/listing/FlowOverride.java
@@ -185,11 +185,7 @@ public enum FlowOverride {
 				if (flowType.isComputed()) {
 					return RefType.CONDITIONAL_COMPUTED_CALL;
 				}
-				if (flowType.isTerminal()) {
-					// assume return was replaced
-					return RefType.COMPUTED_CALL_TERMINATOR;
-				}
-				return flowType;  // don't replace
+				return RefType.CONDITIONAL_CALL_TERMINATOR;
 			}
 			if (flowType.isComputed()) {
 				return RefType.COMPUTED_CALL_TERMINATOR;
-- 
2.45.0

